File: src-tauri/src/handlers/backup.rs
use axum::extract::State;
use axum::Json;

use crate::errors::AppResult;
use crate::handlers::ApiResponse;
use crate::service::backup::{self, BackupConfig, BackupItem};
use crate::service::AppState;

use serde::Deserialize;

pub async fn get_config(State(_s): State<AppState>) -> AppResult<Json<ApiResponse<BackupConfig>>> {
    let cfg = backup::load_config()?;
    Ok(ApiResponse::ok(cfg))
}

#[derive(Deserialize)]
pub struct SaveConfigInput {
    pub endpoint: String,
    pub bucket: String,
    pub region: String,
    pub access_key: String,
    pub secret_key: String,
    pub path_prefix: String,
}

pub async fn save_config(
    _state: State<AppState>,
    Json(input): Json<SaveConfigInput>,
) -> AppResult<Json<ApiResponse<serde_json::Value>>> {
    let cfg = BackupConfig {
        endpoint: input.endpoint,
        bucket: input.bucket,
        region: input.region,
        access_key: input.access_key,
        secret_key: input.secret_key,
        path_prefix: input.path_prefix,
    };
    backup::save_config(&cfg)?;
    Ok(crate::handlers::ok_msg("saved"))
}

#[derive(Deserialize)]
pub struct TestInput {
    pub endpoint: String,
    pub bucket: String,
    pub region: String,
    pub access_key: String,
    pub secret_key: String,
    pub path_prefix: String,
}

pub async fn test_connection(
    _state: State<AppState>,
    Json(input): Json<TestInput>,
) -> AppResult<Json<ApiResponse<serde_json::Value>>> {
    let cfg = BackupConfig {
        endpoint: input.endpoint,
        bucket: input.bucket,
        region: input.region,
        access_key: input.access_key,
        secret_key: input.secret_key,
        path_prefix: input.path_prefix,
    };
    backup::test_connection(&cfg).await?;
    Ok(crate::handlers::ok_msg("ok"))
}

#[derive(Deserialize)]
pub struct CreateBackupInput {
    pub command_history: Vec<String>,
    pub password: String,
}

pub async fn create_backup(
    State(s): State<AppState>,
    Json(input): Json<CreateBackupInput>,
) -> AppResult<Json<ApiResponse<serde_json::Value>>> {
    let cfg = backup::load_config()?;
    let key = backup::create_backup(
        &s.db,
        &s.config.crypto_key,
        &cfg,
        input.command_history,
        input.password,
    )
    .await?;
    Ok(Json(ApiResponse {
        code: 0,
        message: "ok".into(),
        data: Some(serde_json::json!({ "key": key })),
    }))
}

pub async fn export_backup(
    State(s): State<AppState>,
    Json(input): Json<CreateBackupInput>,
) -> AppResult<Json<ApiResponse<serde_json::Value>>> {
    let content =
        backup::export_backup(&s.db, &s.config.crypto_key, input.command_history, input.password)
            .await?;
    Ok(Json(ApiResponse {
        code: 0,
        message: "ok".into(),
        data: Some(serde_json::json!({ "content": content })),
    }))
}

pub async fn list_backups(
    _state: State<AppState>,
) -> AppResult<Json<ApiResponse<Vec<BackupItem>>>> {
    let cfg = backup::load_config()?;
    let items = backup::list_backups(&cfg).await?;
    Ok(ApiResponse::ok(items))
}

#[derive(Deserialize)]
pub struct RestoreInput {
    pub key: String,
    pub password: String,
}

pub async fn restore_backup(
    State(s): State<AppState>,
    Json(input): Json<RestoreInput>,
) -> AppResult<Json<ApiResponse<serde_json::Value>>> {
    let cfg = backup::load_config()?;
    let command_history =
        backup::restore_backup(&s.db, &s.config.crypto_key, &cfg, &input.key, &input.password)
            .await?;
    Ok(Json(ApiResponse {
        code: 0,
        message: "ok".into(),
        data: Some(serde_json::json!({ "command_history": command_history })),
    }))
}

#[derive(Deserialize)]
pub struct ImportInput {
    pub content: String,
    pub password: String,
}

pub async fn import_backup(
    State(s): State<AppState>,
    Json(input): Json<ImportInput>,
) -> AppResult<Json<ApiResponse<serde_json::Value>>> {
    let command_history =
        backup::import_backup(&s.db, &s.config.crypto_key, &input.content, &input.password)
            .await?;
    Ok(Json(ApiResponse {
        code: 0,
        message: "ok".into(),
        data: Some(serde_json::json!({ "command_history": command_history })),
    }))
}

#[derive(Deserialize)]
pub struct DeleteInput {
    pub key: String,
}

pub async fn delete_backup(
    _state: State<AppState>,
    Json(input): Json<DeleteInput>,
) -> AppResult<Json<ApiResponse<serde_json::Value>>> {
    let cfg = backup::load_config()?;
    backup::delete_backup(&cfg, &input.key).await?;
    Ok(crate::handlers::ok_msg("deleted"))
}

File: src/api/backup.ts
import { request } from './client'

export interface BackupConfig {
  endpoint: string
  bucket: string
  region: string
  access_key: string
  secret_key: string
  path_prefix: string
}

export interface BackupItem {
  key: string
  timestamp: string
  size: number
}

export interface SaveConfigInput {
  endpoint: string
  bucket: string
  region: string
  access_key: string
  secret_key: string
  path_prefix: string
}

export function getBackupConfig(): Promise<BackupConfig> {
  return request<BackupConfig>('/api/backup/config')
}

export function saveBackupConfig(input: SaveConfigInput): Promise<void> {
  return request<void>('/api/backup/config', { method: 'PUT', json: input })
}

export function testBackupConnection(input: SaveConfigInput): Promise<void> {
  return request<void>('/api/backup/test', { method: 'POST', json: input })
}

export function createBackup(commandHistory: string[], password: string): Promise<{ key: string }> {
  return request<{ key: string }>('/api/backup/create', {
    method: 'POST',
    json: { command_history: commandHistory, password },
  })
}

export function exportBackup(commandHistory: string[], password: string): Promise<{ content: string }> {
  return request<{ content: string }>('/api/backup/export', {
    method: 'POST',
    json: { command_history: commandHistory, password },
  })
}

export function listBackups(): Promise<BackupItem[]> {
  return request<BackupItem[]>('/api/backup/list')
}

export function restoreBackup(key: string, password: string): Promise<{ command_history: string[] }> {
  return request<{ command_history: string[] }>('/api/backup/restore', {
    method: 'POST',
    json: { key, password },
  })
}

export function importBackup(content: string, password: string): Promise<{ command_history: string[] }> {
  return request<{ command_history: string[] }>('/api/backup/import', {
    method: 'POST',
    json: { content, password },
  })
}

export function deleteBackup(key: string): Promise<void> {
  return request<void>('/api/backup/delete', {
    method: 'POST',
    json: { key },
  })
}
